skyportal/handlers/api/shift.py

```diff
@@ -515,3 +517,144 @@ def delete(self, shift_id, user_id):
+            if start_date.date() > datetime.datetime.today().date():
+                return self.error("Please provide start_date < today")
+            # if there is more than 4 weeks, we return an error
+            if (end_date - start_date).days > 28:
+                return self.error("Please provide a period of less than 4 weeks")
+
+        report = {}
+        if start_date and end_date:
+            s = (
+                Shift.query_records_accessible_by(
+                    self.current_user,
+                    mode="read",
+                    options=[
+                        joinedload(Shift.shift_users),
+                    ],
+                )
+                .filter(Shift.start_date >= start_date)
+                .filter(Shift.start_date <= end_date)
+                .order_by(Shift.start_date.asc())
+                .all()
+            )
+        else:
+            s = (
+                Shift.query_records_accessible_by(
+                    self.current_user,
+                    mode="read",
+                    options=[
+                        joinedload(Shift.shift_users),
+                    ],
+                )
+                .filter(Shift.id == shift_id)
+                .all()
+            )
+        if len(s) == 0:
+            return self.error("No shifts found")
+        shifts = []
+        for shift in s:
+            susers = []
+            for su in shift.shift_users:
+                user = su.user.to_dict()
+                user["admin"] = su.admin
+                user["needs_replacement"] = su.needs_replacement
+                del user["oauth_uid"]
+                susers.append(user)
+            shift = shift.to_dict()
+            shift["shift_users"] = susers
+            shifts.append(shift)
+
+        report['shifts'] = {}
+        report['shifts']['total'] = len(shifts)
+        report['shifts']['data'] = shifts
+        if shift_id and not start_date and not end_date:
+            start_date = shifts[0]['start_date']
+            end_date = shifts[0]['end_date']
+
+        gcns = (
+            GcnEvent.query_records_accessible_by(
+                self.current_user,
+                mode="read",
+            )
+            .filter(GcnEvent.dateobs >= start_date)
+            .filter(GcnEvent.dateobs <= end_date)
+            .order_by(GcnEvent.dateobs.asc())
+            .all()
+        )
+        gcn_added_during_shifts = []
+        # get the gcns added during the shifts
+        if len(gcns) > 0:
+            for shift in shifts:
+                # get list of user_id who are shift_users
+                for gcn in gcns:
+                    gcn = gcn.to_dict()
+                    if (
+                        gcn["dateobs"] >= shift["start_date"]
+                        and gcn["dateobs"] <= shift["end_date"]
+                    ):
+                        if gcn["id"] not in [
+                            gcn["id"] for gcn in gcn_added_during_shifts
```

Comment:
I'd prefer the inner variable to have a different name in the list comprehension. I would even use `g` if it is local to that one line.